### PR TITLE
Add metrics for mobile versions snapshots

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
As a follow up of https://github.com/mattermost/mattermost/pull/27570 we add a new metric to take "snapshots" of the mobile versions currently being used on the platform. This can get us graphs like this one:

Screenshot from 2024-09-13 18-38-34

IMPORTANT: Right now there are two label groups: version (taken from the prop) and platform (taken from the deviceID), and we count ALL sessions. So far it is done this way for a couple of reasons:
- Traditionally, we define a "mobile session" because it has a deviceID. But the true is that many sessions may not have a device id (because it is a custom build that is not signed, or it is signed for a different push proxy and has been removed). So we will be able to find sessions with no deviceID and a mobile version.
- Old phone versions will not inform of the version, so we shouldn't filter by "a version number exist"

So... the way things are done now we can see:
- The number of total sessions
- The approximate number of total sessions (old phones without device ID won't be seen as mobile sessions)
- The approximate number of platform sessions (platform is taken from the device ID, so no device ID, no platform)
- The number of sessions by platform, starting on 2.21 (I think that is the target release for the mobile change)

We can play on how this behave, or we could even add the platform to the version to make things "easier". Let me know what you think.

#### Ticket Link
NONE

#### Release Note
```release-note
Add metric to prometheus to check the mobile versions for each session daily
```

### CODE REVIEW

File: server/channels/jobs/mobile_versions/scheduler.go
Comment (marianunez) [inline server/channels/jobs/mobile_versions/scheduler.go:19]: Is it not possible to reuse the current daily job that sends daily server telemetry to Rudder so we don't have to add a new one?

https://github.com/mattermost/mattermost/blob/6cece9faa6eae10b97e319e8cdeab5477708a570/server/platform/services/telemetry/telemetry.go#L1389

File: server/channels/jobs/mobile_versions/worker.go
Comment (marianunez) [inline server/channels/jobs/mobile_versions/worker.go:28]: How about the other prop we started collecting as well for disabled notifications on device? Could we include that as well?

File: server/channels/jobs/mobile_versions/scheduler.go
Comment (larkox) [reply server/channels/jobs/mobile_versions/scheduler.go:19]: The telemetry job is encapsulated into an interface that doesn't have (apparently) access to the metrics interface.

File: server/channels/jobs/mobile_versions/worker.go
Comment (larkox) [reply server/channels/jobs/mobile_versions/worker.go:28]: I should be doable. I will add it.

File: server/channels/store/sqlstore/session_store.go
Comment (devinbinnie) [inline server/channels/store/sqlstore/session_store.go:170]: Did we check the query plan for this query? We have a lot of functions being done within SQL, as well as no filtering on the sessions table which for some customers might be really really large.

Since this is a job it likely won't be disruptive, however we should see how this performs on large data sets anyways.

**Review by devinbinnie — APPROVED** [review]
Looks good, I have one comment we should consider before merging, but I won't block on that.

File: server/public/model/session.go
Comment (marianunez) [inline server/public/model/session.go:47]: Do we want the `NotificationDisabled` associated to the version though? It seems this would duplicate cardinality. We should be good to just have a total count of disabled ones regardless of version

File: server/channels/jobs/mobile_versions/scheduler.go
Comment (larkox) [reply server/channels/jobs/mobile_versions/scheduler.go:19]: Yes, it should.

File: server/public/model/session.go
Comment (larkox) [reply server/public/model/session.go:47]: We have several options here.

If we look into prometheus side, we can have two different metrics:
- Mobile versions / platform
- Notifications disabled / platform

If we do that instead of `X * 3 * 2` we have `(X * 3) + (2 * 3)` which should be better in terms of cardinality. Now... there are two ways to achieve that.
- Get the data as we are now, and compose it in Golang
- Make two database requests, one for each metric

In this case I have the gut feeling that it is better to get everything, and just compose the values in Golang.

All that being said... we only store one set of values every day, so not sure how "dangerous" is the cardinality in this situation. So right now I lean towards keeping the code simple. But if you have a strong opinion about reducing the cardinality, I can go with 2 metrics and composition in Golang.

Comment (marianunez) [reply server/public/model/session.go]: You are right, I was thinking more of the problem of cardinality with labels but forget we are not storing this as labels of other values so I'm good with your original approach 👍

File: server/channels/store/sqlstore/session_store.go
Comment (larkox) [reply server/channels/store/sqlstore/session_store.go:170]: I checked in my local with 30k sessions, and I got this query plan:

```
                                                                                                QUERY PLAN                                                                                                
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=6323.64..7225.68 rows=30068 width=104)
   Group Key: (COALESCE((props ->> 'mobile_version'::text), 'N/A'::text)), (split_part((deviceid)::text, ':'::text, 1)), (COALESCE((props ->> 'device_notification_disabled'::text), 'false'::text))
   ->  Sort  (cost=6323.64..6398.81 rows=30068 width=123)
         Sort Key: (COALESCE((props ->> 'mobile_version'::text), 'N/A'::text)), (split_part((deviceid)::text, ':'::text, 1)), (COALESCE((props ->> 'device_notification_disabled'::text), 'false'::text))
         ->  Seq Scan on sessions  (cost=0.00..2134.19 rows=30068 width=123)
```

The execution of the query itself seemed immediate in my local. The results (mostly generated randomly) looked like this:
```
 count |    platform     | version | notificationdisabled 
-------+-----------------+---------+----------------------
  1343 |                 | 2.20.0  | false
   648 |                 | 2.20.0  | true
  1313 | android_rn-v2   | 2.20.0  | false
   649 | android_rn-v2   | 2.20.0  | true
  1329 | apple_rnbeta-v2 | 2.20.0  | false
   699 | apple_rnbeta-v2 | 2.20.0  | true
  1312 |                 | 2.21.0  | false
   711 |                 | 2.21.0  | true
  1326 | android_rn-v2   | 2.21.0  | false
   676 | android_rn-v2   | 2.21.0  | true
  1306 | apple_rnbeta-v2 | 2.21.0  | false
   653 | apple_rnbeta-v2 | 2.21.0  | true
  1352 |                 | 2.21.1  | false
   677 |                 | 2.21.1  | true
  1313 | android_rn-v2   | 2.21.1  | false
   662 | android_rn-v2   | 2.21.1  | true
  1378 | apple_rnbeta-v2 | 2.21.1  | false
   637 | apple_rnbeta-v2 | 2.21.1  | true
  1300 |                 | 2.22.0  | false
   668 |                 | 2.22.0  | true
  1359 | android_rn-v2   | 2.22.0  | false
   660 | android_rn-v2   | 2.22.0  | true
  1375 | apple_rnbeta-v2 | 2.22.0  | false
   639 | apple_rnbeta-v2 | 2.22.0  | true
  1949 |                 | N/A     | false
  2044 | android_rn-v2   | N/A     | false
  2023 | apple_rnbeta-v2 | N/A     | false
(27 rows)
```

Comment (devinbinnie) [reply server/channels/store/sqlstore/session_store.go:170]: Okay looks good, we're not seeing any big "table scan" tasks which are usually the slow ones. If it's running quickly on your machine we should be okay. We can double check in load testing if need be.

File: server/channels/app/server.go
Comment (marianunez) [inline server/channels/app/server.go:1570]:
```suggestion
		model.JobTypeMobileSessionMetadata,
```

File: server/channels/jobs/mobile_session_metadata/worker.go
Comment (marianunez) [inline server/channels/jobs/mobile_session_metadata/worker.go:32]: Do we want to log any errors in this cases? Seems that the parent doesn't look at the returned `err`

Comment (larkox) [reply server/channels/jobs/mobile_session_metadata/worker.go:32]: It is not clear where the log is created, but the "parent" logs the error.

The idea is that this will create a worker with this `execute` function. The worker will be passed around, and eventually, the method `Run` of `SimpleWorker` will be called, which eventually calls `DoJob`, and this is where `execute` is really called (`channels/jobs/base_workers.go` line 94). There the error is logged.

**Comment by larkox** [comment]
@marianunez These are a couple of dashboards we could create from this data.

Screenshot from 2024-09-18 12-36-36

**Comment by marianunez** [comment]
> @marianunez These are a couple of dashboards we could create from this data.
> 
> Screenshot from 2024-09-18 12-37-40

Seems we have all the data that can be helpful to understand the ack rate 👍  Later we can see how we can represent it in the dashboard a bit more condensed in the hopes of adding it to the existing notifications one.

File: server/channels/jobs/mobile_session_metadata/worker.go
Comment (marianunez) [reply server/channels/jobs/mobile_session_metadata/worker.go:32]: Thanks for clarifying, I'm not too familiar with how the jobs are setup 👍

**Review by marianunez — APPROVED** [review]

**Comment by agnivade** [comment]
@cpoile - Could you give this a quick look? It adds a new job, and I wanted to run by you if you have any "best practices" for Daniel to follow.

**Review by cpoile — APPROVED** [review]
Looks good -- luckily the simple worker is pretty straightforward. Nice work @larkox!

**Comment by larkox** [comment]
Skipping QA review since this would need a server running for a full day in order to start sending the metrics to grafana. We can test how it behaves with community.

**Pull request merged**
